### ISSUE TITLE
Improve performance when used in rope libraries

### ISSUE THREAD

**Issue by josephg** [cessen/str_indices#9]
Hey! I'm using the str_utils code in another project and I was reading the sse code to understand it better.

In ByteChunk for sse2::__m128i the code currently does this:

```rust
    #[inline(always)]
    fn sum_bytes(&self) -> usize {
        const ONES: u64 = std::u64::MAX / 0xFF;
        let tmp = unsafe { std::mem::transmute::<Self, (u64, u64)>(*self) };
        let a = tmp.0.wrapping_mul(ONES) >> (7 * 8);
        let b = tmp.1.wrapping_mul(ONES) >> (7 * 8);
        (a + b) as usize
    }
```

.. Which is a neat trick, but it makes the "vertical" loop have to build the accumulator every 31 iterations so you don't overflow. I'm no expert at this stuff, but [some reading](https://stackoverflow.com/questions/6996764/fastest-way-to-do-horizontal-sse-vector-sum-or-other-reduction) recommended using [`PSADBW(x, 0)`](https://www.felixcloutier.com/x86/psadbw) ("Compute sum of absolute differences") instead to accumulate into the array.

So changing the code to this:

```rust
    #[inline(always)]
    fn max_acc() -> usize {
        255
    }

    #[inline(always)]
    fn sum_bytes(&self) -> usize {
        unsafe {
            let zero = sse2::_mm_setzero_si128();
            let diff = sse2::_mm_sad_epu8(*self, zero);
            let (low, high) = std::mem::transmute::<Self, (u64, u64)>(diff);
            (low + high) as usize
        }
    }
```

This yields a (modest) performance improvement on my ryzen 5800:

```
ropey:master $ taskset 0x1 nice -10 RUSTFLAGS=-C target-cpu=native cargo criterion -- --measurement-time=10 index_convert
   Compiling ropey v1.3.2 (/home/seph/3rdparty/ropey)
    Finished bench [optimized] target(s) in 37.01s
index_convert/byte_to_char                                                                             
                        time:   [41.762 ns 41.799 ns 41.837 ns]
                        change: [-1.0722% -0.8577% -0.6697%] (p = 0.00 < 0.05)
                        Change within noise threshold.
index_convert/byte_to_line                                                                            
                        time:   [103.24 ns 103.25 ns 103.27 ns]
                        change: [+1.1863% +1.2842% +1.3631%] (p = 0.00 < 0.05)
                        Performance has regressed.
index_convert/char_to_byte                                                                             
                        time:   [87.674 ns 87.701 ns 87.730 ns]
                        change: [-1.6249% -1.5190% -1.4211%] (p = 0.00 < 0.05)
                        Performance has improved.
index_convert/char_to_line                                                                            
                        time:   [153.53 ns 153.55 ns 153.57 ns]
                        change: [-1.4996% -1.3924% -1.2970%] (p = 0.00 < 0.05)
                        Performance has improved.
index_convert/line_to_byte                                                                            
                        time:   [143.57 ns 143.65 ns 143.77 ns]
                        change: [-7.6773% -7.5422% -7.3956%] (p = 0.00 < 0.05)
                        Performance has improved.
index_convert/line_to_char                                                                            
                        time:   [143.31 ns 143.34 ns 143.39 ns]
                        change: [-7.9232% -7.8228% -7.7185%] (p = 0.00 < 0.05)
                        Performance has improved.
```

Is this code change correct?

**Comment by cessen** [comment]
Closing in favor of https://github.com/cessen/str_indices/issues/2

**Comment by cessen** [comment]
Transferred the issue to `str_indices` since it's no longer directly relevant to Ropey.

And re-opening to track progress on this.

**Comment by cessen** [comment]
@josephg
I did benchmarking on my system using the command you gave, but I got essentially the opposite result that you did: using the `str_indices` master-branch functions directly gives a notable performance boost.

Here's the current jumprope-rs master:

```
testdata/direct/automerge-paper                                                                             
                        time:   [9.8263 ms 9.8480 ms 9.8699 ms]
                        thrpt:  [26.320 Melem/s 26.379 Melem/s 26.437 Melem/s]
testdata/buffered/automerge-paper                                                                            
                        time:   [929.02 us 929.56 us 930.19 us]
                        thrpt:  [279.28 Melem/s 279.46 Melem/s 279.63 Melem/s]ed.

Benchmarking testdata/direct/rustcode: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 10.0s. You may wish to increase target time to 10.7s, enable flat sampling, or reduce sample count to 60.
testdata/direct/rustcode                                                                             
                        time:   [2.1165 ms 2.1183 ms 2.1200 ms]
                        thrpt:  [462.20 Melem/s 462.58 Melem/s 462.97 Melem/s]
testdata/buffered/rustcode                                                                            
                        time:   [1.1869 ms 1.1875 ms 1.1880 ms]
                        thrpt:  [824.78 Melem/s 825.16 Melem/s 825.56 Melem/s]

testdata/direct/sveltecomponent                                                                            
                        time:   [822.92 us 823.99 us 825.54 us]
                        thrpt:  [205.34 Melem/s 205.73 Melem/s 205.99 Melem/s]
testdata/buffered/sveltecomponent                                                                            
                        time:   [481.49 us 481.85 us 482.41 us]
                        thrpt:  [351.40 Melem/s 351.80 Melem/s 352.07 Melem/s]d.

testdata/direct/seph-blog1                                                                             
                        time:   [5.8280 ms 5.8386 ms 5.8495 ms]
                        thrpt:  [62.948 Melem/s 63.066 Melem/s 63.181 Melem/s]
testdata/buffered/seph-blog1                                                                            
                        time:   [1.4044 ms 1.4051 ms 1.4058 ms]
                        thrpt:  [261.93 Melem/s 262.07 Melem/s 262.19 Melem/s]
```

And here's using `str_indices` directly ([my diff](https://github.com/cessen/jumprope-rs/commit/18d2c27f39b8c85c1b755821ed37c5bd2ce65425)):

```
testdata/direct/automerge-paper                                                                             
                        time:   [8.8974 ms 8.9312 ms 8.9675 ms]
                        thrpt:  [28.969 Melem/s 29.087 Melem/s 29.197 Melem/s]
                 change:
                        time:   [-9.7231% -9.3096% -8.8700%] (p = 0.00 < 0.05)
                        thrpt:  [+9.7334% +10.265% +10.770%]
                        Performance has improved.
testdata/buffered/automerge-paper                                                                            
                        time:   [891.94 us 892.97 us 894.05 us]
                        thrpt:  [290.56 Melem/s 290.91 Melem/s 291.25 Melem/s]
                 change:
                        time:   [-3.8015% -3.6849% -3.5660%] (p = 0.00 < 0.05)
                        thrpt:  [+3.6978% +3.8259% +3.9518%]
                        Performance has improved.

testdata/direct/rustcode                                                                            
                        time:   [1.9402 ms 1.9472 ms 1.9544 ms]
                        thrpt:  [501.37 Melem/s 503.23 Melem/s 505.02 Melem/s]
                 change:
                        time:   [-8.0181% -7.6477% -7.2564%] (p = 0.00 < 0.05)
                        thrpt:  [+7.8242% +8.2810% +8.7171%]
                        Performance has improved.
testdata/buffered/rustcode                                                                            
                        time:   [1.1383 ms 1.1388 ms 1.1393 ms]
                        thrpt:  [860.07 Melem/s 860.44 Melem/s 860.82 Melem/s]
                 change:
                        time:   [-4.1144% -4.0377% -3.9584%] (p = 0.00 < 0.05)
                        thrpt:  [+4.1216% +4.2076% +4.2909%]
                        Performance has improved.

testdata/direct/sveltecomponent                                                                            
                        time:   [746.23 us 746.79 us 747.37 us]
                        thrpt:  [226.82 Melem/s 227.00 Melem/s 227.16 Melem/s]
                 change:
                        time:   [-9.4873% -9.3365% -9.2025%] (p = 0.00 < 0.05)
                        thrpt:  [+10.135% +10.298% +10.482%]
                        Performance has improved.
testdata/buffered/sveltecomponent                                                                            
                        time:   [460.76 us 460.92 us 461.09 us]
                        thrpt:  [367.65 Melem/s 367.78 Melem/s 367.91 Melem/s]
                 change:
                        time:   [-4.4930% -4.3499% -4.1747%] (p = 0.00 < 0.05)
                        thrpt:  [+4.3566% +4.5477% +4.7044%]
                        Performance has improved.

testdata/direct/seph-blog1                                                                             
                        time:   [5.2779 ms 5.2894 ms 5.3010 ms]
                        thrpt:  [69.462 Melem/s 69.615 Melem/s 69.766 Melem/s]
                 change:
                        time:   [-9.6640% -9.4070% -9.1410%] (p = 0.00 < 0.05)
                        thrpt:  [+10.061% +10.384% +10.698%]
                        Performance has improved.
testdata/buffered/seph-blog1                                                                            
                        time:   [1.3486 ms 1.3502 ms 1.3520 ms]
                        thrpt:  [272.35 Melem/s 272.71 Melem/s 273.03 Melem/s]
                 change:
                        time:   [-3.9445% -3.8368% -3.7281%] (p = 0.00 < 0.05)
                        thrpt:  [+3.8725% +3.9899% +4.1065%]
                        Performance has improved.
```

This on an AMD Threadripper 3960X system on Linux (NixOS), using the exact same benchmarking command you gave.

This is a bit unexpected since I basically just copy-pasted your optimization into `str_utils`, so I assumed at best it would just break even with your code.  But perhaps by using the functions directly, rather than wrapping them like you were doing, the compiler is able to discover some additional optimizations.  Dunno.

I'm curious how [my diff](https://github.com/cessen/jumprope-rs/commit/18d2c27f39b8c85c1b755821ed37c5bd2ce65425) performs on your system.

**Comment by cessen** [comment]
Hmm.  So the mystery deepens.  I reran the benchmarks on my system doing nothing except using str_indices master (no changes to jumprope-rs's code), and I got essentially identical performance improvements.

**Comment by josephg** [comment]
I think you need to be a bit careful swapping between master and 0.3.1 in cargo because the Cargo.lock file let it lock in / cache the version from git and use that even after you change back to the version in cargo. I had to shake it a bit to make it use the cargo version again.

It could also be one of these situations where subtle binary ordering misaligns jumps or something; and there’s no difference between the versions except magic compiler nonsense. And the optimizer can’t run some of its passes across the crate boundary because each crate is compiled separately.

**Comment by cessen** [comment]
> I think you need to be a bit careful swapping between master and 0.3.1 in cargo because the Cargo.lock file let it lock in / cache the version from git and use that even after you change back to the version in cargo.

I don't think that's what's going on here.  I get the same perf difference even if I do `cargo clean` and `rm Cargo.lock` between benchmark runs.

> It could also be one of these situations where subtle binary ordering misaligns jumps or something.

That's possible.  I'm not sure how to investigate that, though.

> And the optimizer can’t run some of its passes across the crate boundary because each crate is compiled separately.

I don't *think* that's a factor here.  All the public functions in str_indices are `#[inline]`, which lets rustc inline across crate boundaries, even in standard non-LTO builds.  And all of the private functions are `#[inline(always)]`, so they should always get inlined into the public functions (unless something really funky is going on).  And on top of that, your release build for jumprope-rs has LTO enabled.

I also want to emphasize that in my runs there doesn't seem to be any perf difference between jumprope-rs master + str_indices master vs jumprope-rs with the diff I linked to:

```
testdata/direct/automerge-paper                                                                             
                        time:   [8.9489 ms 8.9736 ms 8.9987 ms]
                        thrpt:  [28.868 Melem/s 28.949 Melem/s 29.029 Melem/s]
                 change:
                        time:   [-0.4514% -0.1052% +0.2688%] (p = 0.58 > 0.05)
                        thrpt:  [-0.2680% +0.1054% +0.4535%]
                        No change in performance detected.
testdata/buffered/automerge-paper                                                                            
                        time:   [888.86 us 889.21 us 889.58 us]
                        thrpt:  [292.02 Melem/s 292.14 Melem/s 292.26 Melem/s]
                 change:
                        time:   [-0.7300% -0.6488% -0.5676%] (p = 0.00 < 0.05)
                        thrpt:  [+0.5708% +0.6530% +0.7354%]
                        Change within noise threshold.

testdata/direct/rustcode                                                                            
                        time:   [1.9248 ms 1.9265 ms 1.9282 ms]
                        thrpt:  [508.18 Melem/s 508.63 Melem/s 509.07 Melem/s]
                 change:
                        time:   [-0.0168% +0.1911% +0.3882%] (p = 0.07 > 0.05)
                        thrpt:  [-0.3867% -0.1907% +0.0168%]
                        No change in performance detected.
testdata/buffered/rustcode                                                                            
                        time:   [1.1388 ms 1.1393 ms 1.1399 ms]
                        thrpt:  [859.63 Melem/s 860.03 Melem/s 860.40 Melem/s]
                 change:
                        time:   [+0.1183% +0.2181% +0.3155%] (p = 0.00 < 0.05)
                        thrpt:  [-0.3145% -0.2176% -0.1182%]
                        Change within noise threshold.

testdata/direct/sveltecomponent                                                                            
                        time:   [744.97 us 745.77 us 746.66 us]
                        thrpt:  [227.03 Melem/s 227.31 Melem/s 227.55 Melem/s]
                 change:
                        time:   [-0.0365% +0.1529% +0.3374%] (p = 0.13 > 0.05)
                        thrpt:  [-0.3363% -0.1526% +0.0366%]
                        No change in performance detected.
testdata/buffered/sveltecomponent                                                                            
                        time:   [459.70 us 459.91 us 460.13 us]
                        thrpt:  [368.41 Melem/s 368.59 Melem/s 368.75 Melem/s]
                 change:
                        time:   [-0.6885% -0.5794% -0.4833%] (p = 0.00 < 0.05)
                        thrpt:  [+0.4856% +0.5828% +0.6932%]
                        Change within noise threshold.

testdata/direct/seph-blog1                                                                             
                        time:   [5.2613 ms 5.2710 ms 5.2807 ms]
                        thrpt:  [69.728 Melem/s 69.857 Melem/s 69.986 Melem/s]
                 change:
                        time:   [-0.8092% -0.4691% -0.1331%] (p = 0.01 < 0.05)
                        thrpt:  [+0.1333% +0.4713% +0.8158%]
                        Change within noise threshold.
testdata/buffered/seph-blog1                                                                            
                        time:   [1.3469 ms 1.3474 ms 1.3480 ms]
                        thrpt:  [273.16 Melem/s 273.28 Melem/s 273.39 Melem/s]
                 change:
                        time:   [-0.1211% -0.0113% +0.0969%] (p = 0.84 > 0.05)
                        thrpt:  [-0.0968% +0.0113% +0.1213%]
                        No change in performance detected.
```

Which is basically what I would expect with successful inlining, since I ported the same optimization you did to str_indices.

So the situation from my perspective is that both of us are running into unexpected things:
1. In your run, it seems odd that you're getting such a significant perf hit when swapping out your inline optimization with str_indices master (which now has the same optimization).
2. In my run, it seems odd that I'm getting such a significant perf *improvement* with str_indices master without doing anything else, since the only change was more-or-less porting over your optimization (which was already in your code).

So I'm not really sure what to make of all of this yet.  If it was just your case in isolation, I'd say that rustc isn't inlining the str_utils functions.  And that may still be what's going on, although it's then not totally clear why that's not an issue in my runs.

Out of curiosity, what architecture and OS was your benchmark run on?

**Comment by cessen** [comment]
Another data point:

@archseer pointed out that we should probably also be confirming rust versions, in addition to architecture.  Turns out I was still on 1.58.  So I updated to 1.59 and ran the benchmarks again.  Here's the comparative benchmark between jumprope-rs master and jumprope-rs with my diff, using rust 1.59:

```
testdata/direct/automerge-paper                                                                             
                        time:   [9.9392 ms 9.9667 ms 9.9949 ms]
                        thrpt:  [25.991 Melem/s 26.065 Melem/s 26.137 Melem/s]
                 change:
                        time:   [-4.2883% -3.9255% -3.5531%] (p = 0.00 < 0.05)
                        thrpt:  [+3.6840% +4.0859% +4.4804%]
                        Performance has improved.
testdata/buffered/automerge-paper                                                                            
                        time:   [869.45 us 870.23 us 871.02 us]
                        thrpt:  [298.25 Melem/s 298.52 Melem/s 298.78 Melem/s]
                 change:
                        time:   [-5.1343% -4.8745% -4.4811%] (p = 0.00 < 0.05)
                        thrpt:  [+4.6913% +5.1243% +5.4122%]
                        Performance has improved.

Benchmarking testdata/direct/rustcode: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 10.0s. You may wish to increase target time to 10.6s, enable flat sampling, or reduce sample count to 60.
testdata/direct/rustcode                                                                             
                        time:   [2.0995 ms 2.1034 ms 2.1080 ms]
                        thrpt:  [464.82 Melem/s 465.85 Melem/s 466.70 Melem/s]
                 change:
                        time:   [+2.0286% +2.2489% +2.4635%] (p = 0.00 < 0.05)
                        thrpt:  [-2.4043% -2.1994% -1.9883%]
                        Performance has regressed.
testdata/buffered/rustcode                                                                            
                        time:   [1.1142 ms 1.1158 ms 1.1174 ms]
                        thrpt:  [876.95 Melem/s 878.15 Melem/s 879.44 Melem/s]
                 change:
                        time:   [-3.9507% -3.8029% -3.6628%] (p = 0.00 < 0.05)
                        thrpt:  [+3.8020% +3.9533% +4.1132%]
                        Performance has improved.

testdata/direct/sveltecomponent                                                                            
                        time:   [820.95 us 822.71 us 824.55 us]
                        thrpt:  [205.59 Melem/s 206.05 Melem/s 206.49 Melem/s]
                 change:
                        time:   [+1.7673% +2.0638% +2.4178%] (p = 0.00 < 0.05)
                        thrpt:  [-2.3607% -2.0220% -1.7366%]
                        Performance has regressed.
testdata/buffered/sveltecomponent                                                                            
                        time:   [447.29 us 447.96 us 448.61 us]
                        thrpt:  [377.88 Melem/s 378.42 Melem/s 378.99 Melem/s]
                 change:
                        time:   [-5.3759% -5.2243% -5.0393%] (p = 0.00 < 0.05)
                        thrpt:  [+5.3068% +5.5122% +5.6814%]
                        Performance has improved.

testdata/direct/seph-blog1                                                                             
                        time:   [5.7911 ms 5.8102 ms 5.8307 ms]
                        thrpt:  [63.151 Melem/s 63.375 Melem/s 63.583 Melem/s]
                 change:
                        time:   [-1.7391% -1.3685% -0.9661%] (p = 0.00 < 0.05)
                        thrpt:  [+0.9756% +1.3875% +1.7699%]
                        Change within noise threshold.
testdata/buffered/seph-blog1                                                                            
                        time:   [1.3144 ms 1.3164 ms 1.3181 ms]
                        thrpt:  [279.35 Melem/s 279.72 Melem/s 280.13 Melem/s]
                 change:
                        time:   [-3.8583% -3.6996% -3.5186%] (p = 0.00 < 0.05)
                        thrpt:  [+3.6469% +3.8417% +4.0132%]
                        Performance has improved.
```

Still not the same as yours, @josephg, but is nevertheless notably different from 1.58, and does cause some regressions between jumprope-rs master and my diff.  It's also interesting to compare it to the timings of the 1.58 benchmarks, because it looks like 1.59 is notably worse at optimizing in this case.

**Comment by cessen** [comment]
Also benched on nightly, and it's worse than 1.58 across the board.

**Comment by cessen** [comment]
I've been playing around with this more.  When I mark str_indices's functions as `#[inline(always)]`, forcing everything to get inlined, the performance improves to be more-or-less the same as my earlier rust 1.58 results, even on nightly.

So my best guess at this point (emphasis on guess) is that the size of `str_indices::chars::count()` is teetering on the edge of what the compiler considered inlineable, so whether it gets inlined or not varies between compiler versions.

If that's the case, then what's going is that when the early-out optimization is in jumprope-rs's wrapper function, that tiny wrapper function gets inlined and so does the early-out optimization.  But when that same optimization is directly in `str_indices::chars::count()', then the optimization getting inlined depends on whether the compiler decides to inline that whole function or not.

With that theory in mind, I tried splitting the `count_inner()` function into two parts:

1. The part with the early-out optimization.
2. The part with the heavy-weight simd code path, marked as `#[inline(never)]` to prevent it from bloating the calling function.

@josephg 
I've put that experiment in the [encourage_small_inline](https://github.com/cessen/str_indices/tree/encourage_small_inline) branch.  On my machine, it results in jumprope-rs benchmarking results on nightly that are comparable to forcing everything to be inlined.  So if you could test that branch out on your machine as well, I'd be very curious about the results.

However, that branch also results in notably worse benchmarking results for longer (100+ byte) strings in str_indices's own bechmark suite.  So I don't think this is actually a good way forward.

I'm beginning to think that it may be best to just leave this optimization out of str_indices itself.  Then client code can implement it when it's important, as you currently are in jumprope-rs.  Of course, I like the idea of str_indices itself being optimal here, but that may not be possible at the moment in a way that actually gives maximum benefit to the client.

**Comment by josephg** [comment]
Oh interesting! I didn't expect that, and probably wouldn't have tried it. That makes a lot of sense to me. Thanks for investigating!

And yeah, that makes sense. I'm not actually convinced inlining the char count methods is always appropriate. They impact performance, but so do a lot of other things in the codebase! In jumprope I mark if a leaf only has ASCII characters anyway, and skip calling these methods entirely if thats the case. And in wasm projects, code size does matter a lot.

Anyway, I'm happy to leave this in the "fuzzy compiler choices" territory. A 4% performance difference isn't crazy (especially compared to adding that wrapper around the library for buffered writes.) And jumprope currently doesn't work properly under MIRI, which isn't a huge problem but its annoying. I think fixing that should take priority.

Oh and if you want a library like this for text editors, it might *also* make sense to add an index by lines like you have in ropey. Lots of more important things to play around with!

**Comment by cessen** [comment]
Okay, sounds good!

I decided to make a compromise and remove the single-byte-string optimization, but not the more general short-string optimization.  It still gives a nice 2x performance boost for very short strings, and is something I think I can generalize more easily to the other functions in str_indices as well, so that it becomes more of a general pattern in the code base.

In any case, please don't hesitate to file issues here if you run into any other performance hiccups, or optimizations you think might make sense to incorporate into str_indices. :-)

Edit: oh!  And of course, thanks so much for your help on this!

> Oh and if you want a library like this for text editors, it might also make sense to add an index by lines like you have in ropey.

Do you mean jumprope or str_indices?  (str_indices already has line indexing functions, so I assume you mean jumprope?)

**Comment by cessen** [comment]
Also, just published [v0.3.2](https://crates.io/crates/str_indices/0.3.2) with the aforementioned changes.

**Comment by josephg** [comment]
Awesome sounds good!

And yeah I mean in jumprope.
